### PR TITLE
Add support to create or terminate AWS Elastic Map Reduce

### PR DESCRIPTION
**Pull request by wellingh**
##### SUMMARY
- create or terminating EMR Clusters
- steps (jobs spark for example) can be setup when Cluster is creating
- because the name is not a unique field in AWS this module can't ensure the state from a cluster, so if you runs this module 10 times, 10 cluster will be created
- basically this module only receive args and send to boto3, so any argument supported by method run_job_flow() can be used when state is create. Check http://boto3.readthedocs.io/en/latest/reference/services/emr.html#EMR.Client.run_job_flow for more information.
- it depends on boto3

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
emr_cluster

##### ANSIBLE VERSION
```
ansible 2.8.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/wellington.ramos/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/wellington.ramos/git/ansible/lib/ansible
  executable location = /home/wellington.ramos/git/ansible/bin/ansible
  python version = 3.7.2 (default, Mar 17 2019, 16:39:09) [GCC 6.3.0 20170516]
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: State should have the choices 'present' or 'absent' to be like other Ansible modules.

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: It should be a dict, flattened like other AWS Ansible modules. You can use boto3_tag_list_to_ansible_dict for the transformation.

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: Should also handle BotoCoreError

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: This can be updated to AnsibleAWSModule from ansible.module_utils.aws.core

That will also mean you can remove the imports and uses of get_aws_connection_info, boto3_conn, and HAS_BOTO3. You can create the client with `connection = module.client('emr')` (which does not need exception handling around it)

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: This needs BotoCoreError and ClientError exception handling:
```
except (BotoCoreError, ClientError) as e:
    module.fail_json_aws(e, msg="Unable to terminate job flows for cluster {0}".format(cluster_id))
```

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: Needs exception handling here too.

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py:700]: There is a built-in waiter you can use for this: http://boto3.readthedocs.io/en/latest/reference/services/emr.html#EMR.Waiter.ClusterTerminated

You can modify the waiter config to wait the correct amount of time by adding the WaiterConfig keyword:
```
waiter.wait(
    ClusterId=cluster_id,
    WaiterConfig={'Delay': 5, 'MaxAttempts': wait_timeout // 5}
)['Cluster']
```

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: There is a waiter for this too: http://boto3.readthedocs.io/en/latest/reference/services/emr.html#EMR.Waiter.ClusterRunning

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: This needs exception handling.

**Review by s-hertel — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: Rather than link to other docs, you can document complex options.
```
ebs_volumes:
    description: A list of volumes to create
    type: complex
    contains:
        VolumeSpecification:
            description: a dict ...
            type: complex
            contains:
                VolumeType:
                    ...
                Iops:
                    ...
                SizeInGB:
        VolumesPerInstance: ...
```

Then in the argument spec this would look like:
```
ebs_volumes=dict(
    type='list', elements='dict', options=dict(
        VolumeSpecification=dict(type='dict', elements='dict', options=dict(
            VolumeType=dict(....<document like normal options>),
            Iops=dict(....),
            SizeInGB=dict(....),
        )
        VolumesPerInstance=dict(...),
))
```
Here's a module example you can look at: https://github.com/ansible/ansible/blob/858f0fc00006e15d33b63a505c80b0a1412a4a7e/lib/ansible/modules/cloud/amazon/elb_network_lb.py#L368-L375

Another nice thing would be to add aliases to the suboptions, so people can provide VolumesPerInstance as the usual snake_case volumes_per_instance, since that is the typical pattern for Ansible AWS modules.

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: `- default present` isn't the same as `default: present`. The latter gets parsed correctly. You'll fix a lot of the shippable errors by correcting that. All the others too. You should also document choices for the options that have them. So it should look like:
```
state:
  description:
    - Whether the cluster should exist or be absent
  default: present
  choices: ['absent', 'present']
```

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: This should be called tags for consistency with the other AWS modules. It also appears tags are only changed/added by this module when first creating the job flow but it should be modifiable since the API supports this with add_tags() and remove_tags(). Here are some general guidelines for dealing with tags. https://github.com/ansible/ansible/blob/bf304832ffb8fc723c4d2f755e2e45ba17809f73/lib/ansible/modules/cloud/amazon/GUIDELINES.md#dealing-with-tags

Here are a couple module examples, though there are many:
https://github.com/ansible/ansible/blob/ab96a841542478189a6195a05ff748565613056f/lib/ansible/modules/cloud/amazon/ec2_ami.py#L568-L585
https://github.com/ansible/ansible/blob/1905a6e8fb38a6d85eed5632c45250aa676715d9/lib/ansible/modules/cloud/amazon/ec2_vpc_route_table.py#L337-L360

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: `state` shouldn't be an action. This should just have the choices `present` and `absent`.

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: This never gets raised so you can remove it and where it's used in main().

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: This should have {0} rather than {} for Python 2.6 compatibility.

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: This has a paginator that should be used in case there are too many clusters to return with list_clusters() alone. https://boto3.readthedocs.io/en/latest/reference/services/emr.html#EMR.Paginator.ListClusters

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: Rather than fail here, you can use a required_if in AnsibleAWSModule. Same for cluster_id for removing the cluster:
```
... = AnsibleAWSModule(...
                       required_if=[
                           ['state', 'present', ['name']],
                           ['state', 'absent', ['cluster_id']],
                       ]
)
```

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py:713]: What if the cluster has been deleted in the previous playbook run? This should be idempotent. Right now it looks like it will always attempt to delete it, and with most AWS services trying to delete something that isn't there fails.

**Review by s-hertel — CHANGES_REQUESTED** [review]
@wmaramos Thanks for updating some things and using the waiter! Are you able to fix the things the Shippable found (that the bot commented)? Also, as per policy, all new modules after 2.5 need integration tests https://github.com/ansible/community/blob/master/group-aws/integration.md#policy. They don't need to be complicated, but should test creating/modifying/removing a cluster. Here's an example of the integration test structure for the ec2_group module: https://github.com/ansible/ansible/tree/devel/test/integration/targets/ec2_group

If you need any help with anything, feel welcome to ping me on IRC (username shertel).

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py]: This should also be returned as tags for consistency with other AWS modules. And it shouldn't be a list of `[{'Key', name1, 'Value', value1}, {'Key': name2, 'Value': value2},]`, but a dictionary like `{name1: value1, name2: value2}`. You can use the `boto3_tag_list_to_ansible_dict` function from ansible.module_utils.ec2 to make that transformation.

**Review by s-hertel — CHANGES_REQUESTED** [review]

**Comment by anryko** [comment]
@wmaramos are you still working on this?

**Comment by anryko** [comment]
@s-hertel I want to help getting this merged. Considering the fact that initial contributor is not responding, what are my options?

**Comment by wmaramos** [comment]
@anryko @s-hertel sorry guys about delay, in the next weekend I'll get free time to work on it so should I rebase with ansible/devel branch?

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (wellingh) [reply lib/ansible/modules/cloud/amazon/emr_cluster.py]: done

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (wellingh) [reply lib/ansible/modules/cloud/amazon/emr_cluster.py:713]: fixed up

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (wellingh) [reply lib/ansible/modules/cloud/amazon/emr_cluster.py]: done

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (wellingh) [reply lib/ansible/modules/cloud/amazon/emr_cluster.py]: done

File: test/integration/targets/emr_cluster/aliases
Comment (mattclay) [inline test/integration/targets/emr_cluster/aliases]: New tests should not be marked unstable. If the tests are truly unstable they should be fixed before merging.

Comment (wellingh) [reply test/integration/targets/emr_cluster/aliases]: ok I just fixed it now, but I saw the tests are failing because I can't create the default IAM roles used by EMR, https://app.shippable.com/github/ansible/ansible/runs/115793/104/console.

what I've got to do?

**Comment by mattclay** [comment]
@wmaramos This PR has been added to the [backlog](https://github.com/ansible/ansible/labels/needs_ci_update) of PRs blocked on CI updates. Unfortunately I don't know when we'll be able to get to this. Marking the test unsupported in the meantime will allow merging -- at the risk of having future modifications to the module and/or tests potentially result in undetected failures.

**Comment by jeking3** [comment]
@mattclay there are 16 AWS PRs blocked on needing CI update, some like this for over a year.  That seems pretty unreasonable to me.  How do we escalate this issue and get these unblocked?

**Comment by mattclay** [comment]
@jeking3 During PyCon 2019 this week I published https://github.com/mattclay/aws-terminator, which provides the policies and code we use to support the AWS module tests. We already made some good progress during the PyCon sprints to support additional tests, but there's a lot more to be done.

If you're able to contribute there that would be the best way to help unblock PRs needing CI updates.

**Comment by jeking3** [comment]
It's quite a shame Amazon doesn't make it easier to clean up things, you know, like Azure resource groups does.

**Comment by mattclay** [comment]
@jeking3 Agreed. That's one of the nice things about using Azure. However, it's not flawless. We regularly end up with resources which cannot be deleted and need to open support tickets to get them cleaned up. I have one such ticket open right now for 58 public IP addresses.

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (markocelan) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py:383]: _action_on_failure_ should be in the same level as _hadoop_jar_step_

File: lib/ansible/modules/cloud/amazon/emr_cluster.py
Comment (markocelan) [inline lib/ansible/modules/cloud/amazon/emr_cluster.py:727]: This results in following error: 

> RuntimeError: dictionary changed size during iteration

I solved it by doing it in two steps:

```
params_to_delete = set()

for param in params.keys():
        if param not in emr_argument_spec():
            params_to_delete.add(param)

for param in params_to_delete:
        del params[param]
```

**Review by markocelan — CHANGES_REQUESTED** [review]
While testing your module I've stumbled upon two issues.

**Comment by Akasurde** [comment]
Please submit a new module request to https://github.com/ansible-collections/community.aws. Thanks for your contribution.

**Comment by goneri** [comment]
The same request in the community.aws https://github.com/ansible-collections/community.aws/issues/289

**Pull request closed without merging**
